### ISSUE TITLE
Reduce indirect memory-memory copies using an MVC instruction

### ISSUE THREAD

**Issue by fjeremic** [eclipse/omr#205]
When the reference counts permit we should prefer to generate an `MVC` instruction instead of a `L` + `ST` instruction pair for memory copy. The instruction timings suggest they are empirically equivalent if not a slight edge for `MVC` for the non-overlapping case. In addition we save 2-bytes worth of instruction cache (icache) and avoid a nasty scenario where the register allocator (RA) chooses to HPR (high-word general purpose register) upgrade both the load and the store resulting in sub-par performance and double the icache footprint than that of an `MVC`. The specific tree can take the following shape:

```
------------------------------
n624n    (  0)  istorei
n600n    (  2)    ==>new (in &GPR_0163) (highWordZero Unsigned X!=0 X>=0 allocationCanBeRemovedsharedMemory )
n626n    (  1)    iloadi
n613n    (  1)      ==>acall (in &GPR_0213)
------------------------------
Calling genericLoadHelper with -32(2)>32 MEM
------------------------------
n624n    (  0)  istorei
n600n    (  1)    ==>new (in &GPR_0163) (highWordZero Unsigned X!=0 X>=0 allocationCanBeRemovedsharedMemory )
n626n    (  0)    iloadi (in GPR_0261)
n613n    (  0)      ==>acall (in &GPR_0213)
------------------------------

[0x51122FE9F0] L       GPR_0261, 4(&GPR_0213)
[0x51122FEBD8] ST      GPR_0261, 8(&GPR_0163)
```

which is then HPR upgraded to result in the following generated sequence:

```
0x35300230 00000136 [0x5074E0FE08] e3 70 80 04 00 ca                        LFH     HPR7, 4(GPR8) ; HPR Upgraded
0x35300236 0000013c [0x5074E0FD10] e3 70 90 08 00 cb                        STFH    HPR7, 8(GPR9) ; HPR Upgraded
```

This reduction is currently only performed for direct loads and stores of `OMR::Int32` datatype however there is no reason why we cannot handle indirect loads and stores of all datatypes as well.

**Comment by 0xdaryl** [comment]
The code looks OK.  Can you change the commit title so that its less than 50 chars and in the imperative voice?  (sorry, I realized I wasn't enforcing this aspect of the contribution guidelines for earlier commits).

**Comment by fjeremic** [comment]
@0xdaryl fixed in ead043bb77d51bef80bbabf38fa2bfaf93df778a.

**Comment by fjeremic** [comment]
It is also noteworthy to reference _Principles of Operation: Chapter 5. Program Execution_ where it states that for `MVC` "the first operand is accessed in a left-to-right direction, and all bytes accessed within a double word appear to be accessed concurrently as observed by other CPUs." This is an important fact that enables this transformation as concurrent threads will not be able to read partially written objects being copied.
